src/ngraph/pass/constant_propagation.cpp

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "constant_propagation.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/pattern/op/label.hpp"
+#include "ngraph/runtime/reference/reshape.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+template <class T>
+shared_ptr<op::Constant> make_constant_reshape(shared_ptr<op::Constant> constant,
+                                               shared_ptr<op::Reshape> reshape)
+{
+    auto out_shape = reshape->get_shape();
+    auto out_vec = vector<T>(shape_size(out_shape));
+
+    runtime::reference::reshape<T>(constant->get_vector<T>().data(),
+                                   out_vec.data(),
+                                   constant->get_shape(),
+                                   reshape->get_input_order(),
+                                   out_shape);
+
+    return make_shared<op::Constant>(constant->get_element_type(), out_shape, out_vec);
+}
+
+void ngraph::pass::ConstantPropagation::construct_constant_reshape()
+{
+    auto constant_pred = [](shared_ptr<Node> n) {
+        return dynamic_pointer_cast<op::Constant>(n) != nullptr;
+    };
+
+    auto constant_label = make_shared<pattern::op::Label>(element::f32, Shape{2, 4}, constant_pred);
+    auto reshape = make_shared<op::Reshape>(constant_label, AxisVector{0, 1}, Shape{2, 4, 1});
+    auto reshape_label = make_shared<pattern::op::Label>(reshape, nullptr, NodeVector{reshape});
```

Comment:
wouldn't `reshape_label` be the same as `m.get_match_root()`? could we use the latter please if possible?

Reply:
fixed